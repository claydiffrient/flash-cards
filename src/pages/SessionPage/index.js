import React, { Component } from "react";
import { connect } from "react-redux";
import Button from "@instructure/ui-buttons/lib/components/Button";
import Grid, {
  GridCol,
  GridRow
} from "@instructure/ui-layout/lib/components/Grid";
import RangeInput from "@instructure/ui-forms/lib/components/RangeInput";
import Card from "../../Card";
import { bool, arrayOf, shape, string, func } from "prop-types";
import { Link } from "react-router-dom";

import themeable from "@instructure/ui-themeable";
import styles from "./styles.css";
import theme from "./theme.js";

import { saveCard, addCard } from "../../actions/cards";
import { fetchDecks } from "../../actions/decks";

export class SessionPage extends Component {
  static propTypes = {
    isEditing: bool,
    deckId: string,
    decks: arrayOf(
      shape({
        id: string,
        cards: arrayOf(
          shape({
            id: string,
            flipped: bool,
            text: shape({
              front: string,
              back: string
            })
          })
        )
      })
    ),
    fetchDecks: func.isRequired,
    addCard: func,
    saveCard: func
  };

  static defaultProps = {
    isEditing: false,
    decks: []
  };

  constructor(props) {
    super(props);
    this.state = {
      currentlyDisplayedIndex: 0,
      showEditor: false,
      timerValue: 0,
      ...this.getDeckProperties(props.decks)
    };
  }

  componentWillMount() {
    this.props.fetchDecks();
  }

  componentWillReceiveProps(nextProps) {
    if (this.props !== nextProps) {
      const { selectedDeck, cards } = this.getDeckProperties(nextProps.decks);
      let currentlyDisplayedIndex = this.state.currentlyDisplayedIndex;
      if (this.state.addingCard) {
        currentlyDisplayedIndex = selectedDeck.cards.length - 1;
      }
      this.setState({
        selectedDeck,
        cards,
        currentlyDisplayedIndex,
        addingCard: false
      });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (
      prevState.timerValue !== this.state.timerValue ||
      prevState.currentlyDisplayedIndex !== this.state.currentlyDisplayedIndex
    ) {
      if (this.state.timerValue === 0 && this.timerId) {
        clearTimeout(this.timerId);
        this.timerId = undefined;
      } else {
        if (this.state.timerValue === 0) {
          return;
        } else {
          this.timerId = setTimeout(() => {
            if (this.card) {
              this.card.flipCard();
              clearTimeout(this.timerId);
              this.timerId = undefined;
            }
          }, this.state.timerValue * 1000);
        }
      }
    }
  }

  compareCardCounts = (originalDecks, newDecks) => {
    if (newDecks.length === originalDecks.length) {
      const oDLength = originalDecks.reduce(
        (acc, curVal) => acc + curVal.cards.length,
        0
      );
      const nDLength = newDecks.reduce(
        (acc, curVal) => acc + curVal.cards.length,
        0
      );
      return oDLength === nDLength;
    } else {
      return false;
    }
  };

  getDeckProperties = decks => {
    const selectedDeck = decks.find(d => d._id === this.props.deckId);
    if (selectedDeck) {
      return {
        selectedDeck,
        cards: selectedDeck.cards
      };
    } else {
      return {
        selectedDeck: {},
        cards: []
      };
    }
  };

  moveNext = () => {
    let nextVal = this.state.currentlyDisplayedIndex + 1;
    if (nextVal > this.state.cards.length - 1) {
      nextVal = 0;
    }
    this.setState({
      currentlyDisplayedIndex: nextVal
    });
  };

  movePrevious = () => {
    let nextVal = this.state.currentlyDisplayedIndex - 1;
    if (nextVal < 0) {
      nextVal = this.state.cards.length - 1;
    }
    this.setState({
      currentlyDisplayedIndex: nextVal
    });
  };

  toggleEditing = () => {
    this.setState({
      showEditor: !this.state.showEditor
    });
  };

  addCard = () => {
    this.setState(
      {
        addingCard: true
      },
      () => {
        this.props.addCard(this.state.selectedDeck._id, {
          text: { front: "New Card Front", back: "New Card Back" }
        });
      }
    );
  };

  handleSave = card => {
    this.props.saveCard(this.state.selectedDeck._id, card);
    this.toggleEditing();
  };

  handleTimerValueChange = val => {
    this.setState({
      timerValue: parseInt(val, 10)
    });
  };

  render() {
    const card = this.state.cards[this.state.currentlyDisplayedIndex];
    const position = `${this.state.currentlyDisplayedIndex + 1} / ${
      this.state.cards.length
    }`;
    if (card) {
      return (
        <Grid>
          <GridRow>
            <GridCol>
              <Card
                key={card.id}
                {...card}
                handleSave={this.handleSave}
                editMode={this.state.showEditor}
                cardCountPosition={position}
                ref={c => {
                  this.card = c;
                }}
              />
            </GridCol>
          </GridRow>
          <GridRow>
            <GridCol width={3}>
              <Button onClick={this.movePrevious}>Previous</Button>
            </GridCol>
            <GridCol width={this.props.isEditing ? 3 : 6}>
              {this.props.isEditing ? (
                <Button onClick={this.toggleEditing}>Toggle Edit</Button>
              ) : (
                <RangeInput
                  min={5}
                  max={60}
                  step={5}
                  value={this.state.timerValue}
                  onChange={this.handleTimerValueChange}
                  label="Time"
                  layout="inline"
                  size="medium"
                />
              )}
            </GridCol>
            <GridCol>
              {this.props.isEditing && (
                <Button onClick={this.addCard}>Add Card</Button>
              )}
            </GridCol>
            <GridCol width={3}>
              <Button onClick={this.moveNext}>Next</Button>
            </GridCol>
          </GridRow>
          <GridRow>
            <GridCol>
              <Button as={Link} to={`/`}>
                Go Home
              </Button>
            </GridCol>
          </GridRow>
        </Grid>
      );
    } else {
      return (
        <Grid>
          <GridRow>
            <GridCol>
              {this.props.isEditing && (
                <Button onClick={this.addCard}>Add Card</Button>
              )}
            </GridCol>
          </GridRow>
        </Grid>
      );
    }
  }
}

const mapStateToProps = state => {
  return {
    cards: state.cards,
    decks: state.decks
  };
};

const mapDispatchToProps = {
  saveCard,
  addCard,
  fetchDecks
};

export const ConnectedSessionPage = connect(
  mapStateToProps,
  mapDispatchToProps
)(SessionPage);

export default themeable(theme, styles)(ConnectedSessionPage);
