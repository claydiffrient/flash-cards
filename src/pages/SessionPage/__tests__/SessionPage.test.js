import React from "react";
import { shallow, mount } from "enzyme";
import SessionPage from "../index";

it("renders", () => {
  const wrapper = mount(<SessionPage />);
  expect(wrapper).toMatchSnapshot();
});
