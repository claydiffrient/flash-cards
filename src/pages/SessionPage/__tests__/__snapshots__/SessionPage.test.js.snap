// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<SessionPage
  addCard={[Function]}
  cards={
    Array [
      Object {
        "flipped": false,
        "id": 1,
        "text": Object {
          "back": "3",
          "front": "What is the square root of 9?",
        },
      },
      Object {
        "flipped": false,
        "id": 2,
        "text": Object {
          "back": "12",
          "front": "What is the square root of 144?",
        },
      },
    ]
  }
  fetchCards={[Function]}
  isEditing={false}
  saveCard={[Function]}
>
  <Grid
    colSpacing="medium"
    hAlign="start"
    rowSpacing="medium"
    startAt="small"
    vAlign="top"
    visualDebug={false}
  >
    <span
      className="_1USmpIr _3JgELif"
    >
      <GridRow
        colSpacing="medium"
        hAlign="start"
        isLastRow={false}
        key=".0"
        rowSpacing="medium"
        startAt="small"
        vAlign="top"
        visualDebug={false}
      >
        <span
          className="_3P46brR _2fgcv6I _35Ws9jw _2lflNF1 EaLykbQ _2E7xkbD"
        >
          <GridCol
            colSpacing="medium"
            hAlign="start"
            isLastCol={true}
            isLastRow={false}
            key=".0"
            rowSpacing="medium"
            startAt="small"
            textAlign="inherit"
            vAlign="top"
            visualDebug={false}
          >
            <span
              className="zlFFAtO _2hZ32yM _1XANqzl _1j7zLHc _2nFm4GG _23makCX dd1PwcH"
            >
              <Card
                editMode={false}
                flipped={false}
                handleSave={[Function]}
                id={1}
                text={
                  Object {
                    "back": "3",
                    "front": "What is the square root of 9?",
                  }
                }
              >
                <div
                  className="Card__card__2rORLZF"
                >
                  <Grid
                    colSpacing="medium"
                    hAlign="start"
                    rowSpacing="medium"
                    startAt="small"
                    vAlign="top"
                    visualDebug={false}
                  >
                    <span
                      className="_1USmpIr _3JgELif"
                    >
                      <GridRow
                        colSpacing="medium"
                        hAlign="start"
                        isLastRow={false}
                        key=".0"
                        rowSpacing="medium"
                        startAt="small"
                        vAlign="top"
                        visualDebug={false}
                      >
                        <span
                          className="_3P46brR _2fgcv6I _35Ws9jw _2lflNF1 EaLykbQ _2E7xkbD"
                        >
                          <GridCol
                            colSpacing="medium"
                            hAlign="start"
                            isLastCol={false}
                            isLastRow={false}
                            key=".0"
                            rowSpacing="medium"
                            startAt="small"
                            textAlign="inherit"
                            vAlign="top"
                            visualDebug={false}
                          >
                            <span
                              className="zlFFAtO _2hZ32yM _1XANqzl _1j7zLHc _2nFm4GG _23makCX"
                            >
                              <Text
                                as="span"
                                letterSpacing="normal"
                                size="medium"
                              >
                                <span
                                  className="_3jLMUSh _3fy1EZS _2WOMSfA"
                                >
                                  <div
                                    dangerouslySetInnerHTML={
                                      Object {
                                        "__html": "What is the square root of 9?",
                                      }
                                    }
                                  />
                                </span>
                              </Text>
                            </span>
                          </GridCol>
                        </span>
                      </GridRow>
                      <GridRow
                        colSpacing="medium"
                        hAlign="start"
                        isLastRow={true}
                        key=".1"
                        rowSpacing="medium"
                        startAt="small"
                        vAlign="top"
                        visualDebug={false}
                      >
                        <span
                          className="_3P46brR M4KzJCD _2fgcv6I _35Ws9jw _2lflNF1 EaLykbQ _2E7xkbD"
                        >
                          <GridCol
                            colSpacing="medium"
                            hAlign="start"
                            isLastCol={true}
                            isLastRow={true}
                            key=".0"
                            rowSpacing="medium"
                            startAt="small"
                            textAlign="inherit"
                            vAlign="top"
                            visualDebug={false}
                          >
                            <span
                              className="zlFFAtO _2hZ32yM _1XANqzl _1j7zLHc _2nFm4GG _23makCX _1X9b06D dd1PwcH"
                            >
                              <Button
                                as="button"
                                buttonRef={[Function]}
                                fluidWidth={false}
                                margin="0"
                                onClick={[Function]}
                                size="medium"
                                type="button"
                                variant="default"
                              >
                                <Container
                                  aria-disabled={null}
                                  as="button"
                                  className="_922amOO _16-3B-L _3Gg19qR"
                                  display={null}
                                  elementRef={[Function]}
                                  margin="0"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="button"
                                  tabIndex="0"
                                  type="button"
                                >
                                  <button
                                    aria-disabled={null}
                                    className="_922amOO _16-3B-L _3Gg19qR"
                                    onClick={[Function]}
                                    onKeyDown={[Function]}
                                    role="button"
                                    style={
                                      Object {
                                        "margin": "0",
                                        "padding": null,
                                      }
                                    }
                                    tabIndex="0"
                                    type="button"
                                  >
                                    <span
                                      className="_2CUXt80"
                                    >
                                      Flip
                                    </span>
                                  </button>
                                </Container>
                              </Button>
                            </span>
                          </GridCol>
                        </span>
                      </GridRow>
                    </span>
                  </Grid>
                </div>
              </Card>
            </span>
          </GridCol>
        </span>
      </GridRow>
      <GridRow
        colSpacing="medium"
        hAlign="start"
        isLastRow={true}
        key=".1"
        rowSpacing="medium"
        startAt="small"
        vAlign="top"
        visualDebug={false}
      >
        <span
          className="_3P46brR M4KzJCD _2fgcv6I _35Ws9jw _2lflNF1 EaLykbQ _2E7xkbD"
        >
          <GridCol
            colSpacing="medium"
            hAlign="start"
            isLastCol={false}
            isLastRow={true}
            key=".0"
            rowSpacing="medium"
            startAt="small"
            textAlign="inherit"
            vAlign="top"
            visualDebug={false}
          >
            <span
              className="zlFFAtO _2hZ32yM _1XANqzl _1j7zLHc _2nFm4GG _23makCX _1X9b06D"
            >
              <Button
                as="button"
                buttonRef={[Function]}
                fluidWidth={false}
                margin="0"
                onClick={[Function]}
                size="medium"
                type="button"
                variant="default"
              >
                <Container
                  aria-disabled={null}
                  as="button"
                  className="_922amOO _16-3B-L _3Gg19qR"
                  display={null}
                  elementRef={[Function]}
                  margin="0"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                  tabIndex="0"
                  type="button"
                >
                  <button
                    aria-disabled={null}
                    className="_922amOO _16-3B-L _3Gg19qR"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="button"
                    style={
                      Object {
                        "margin": "0",
                        "padding": null,
                      }
                    }
                    tabIndex="0"
                    type="button"
                  >
                    <span
                      className="_2CUXt80"
                    >
                      Previous
                    </span>
                  </button>
                </Container>
              </Button>
            </span>
          </GridCol>
          <GridCol
            colSpacing="medium"
            hAlign="start"
            isLastCol={false}
            isLastRow={true}
            key=".1"
            rowSpacing="medium"
            startAt="small"
            textAlign="inherit"
            vAlign="top"
            visualDebug={false}
          >
            <span
              className="zlFFAtO _2hZ32yM _1XANqzl _1j7zLHc _2nFm4GG _23makCX _1X9b06D"
            />
          </GridCol>
          <GridCol
            colSpacing="medium"
            hAlign="start"
            isLastCol={false}
            isLastRow={true}
            key=".2"
            rowSpacing="medium"
            startAt="small"
            textAlign="inherit"
            vAlign="top"
            visualDebug={false}
          >
            <span
              className="zlFFAtO _2hZ32yM _1XANqzl _1j7zLHc _2nFm4GG _23makCX _1X9b06D"
            />
          </GridCol>
          <GridCol
            colSpacing="medium"
            hAlign="start"
            isLastCol={true}
            isLastRow={true}
            key=".3"
            rowSpacing="medium"
            startAt="small"
            textAlign="inherit"
            vAlign="top"
            visualDebug={false}
          >
            <span
              className="zlFFAtO _2hZ32yM _1XANqzl _1j7zLHc _2nFm4GG _23makCX _1X9b06D dd1PwcH"
            >
              <Button
                as="button"
                buttonRef={[Function]}
                fluidWidth={false}
                margin="0"
                onClick={[Function]}
                size="medium"
                type="button"
                variant="default"
              >
                <Container
                  aria-disabled={null}
                  as="button"
                  className="_922amOO _16-3B-L _3Gg19qR"
                  display={null}
                  elementRef={[Function]}
                  margin="0"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                  tabIndex="0"
                  type="button"
                >
                  <button
                    aria-disabled={null}
                    className="_922amOO _16-3B-L _3Gg19qR"
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="button"
                    style={
                      Object {
                        "margin": "0",
                        "padding": null,
                      }
                    }
                    tabIndex="0"
                    type="button"
                  >
                    <span
                      className="_2CUXt80"
                    >
                      Next
                    </span>
                  </button>
                </Container>
              </Button>
            </span>
          </GridCol>
        </span>
      </GridRow>
    </span>
  </Grid>
</SessionPage>
`;
